pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/mergerollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,691 @@
+ *          A new ZNode will be created, with watermarkMs as the smallest time found in all segments truncated to the
+ *          closest bucket start time.
+ *        - The execution window for the task is calculated as,
+ *          bucketStartMs = watermarkMs
+ *          bucketEndMs = bucketStartMs + bucketTimeMs
+ *          - bucketEndMs must be equal or older than the bufferTimeMs
+ *          - bucketEndMs of higher merge level should be less or equal to the waterMarkMs of lower level
+ *        - Bump up target window and watermark if needed.
+ *          - If there's no unmerged segments (by checking segment zk metadata {mergeRollupTask.mergeLevel: level}) for
+ *            current window, keep bumping up the watermark and target window until unmerged segments are found.
+ *          - Else skip the scheduling.
+ *      - Select segments for the bucket:
+ *        - Skip buckets which all segments are merged
+ *        - If there's no spilled over segments (segments spanning multiple time buckets), schedule buckets in parallel
+ *        - Else, schedule buckets till the first one that has spilled over data (included), so the spilled over data
+ *          will be merged next round
+ *      - Create the tasks for the current bucket (and per partition for partitioned tables) based on
+ *        maxNumRecordsPerTask
+ */
+@TaskGenerator
+public class MergeRollupTaskGenerator implements PinotTaskGenerator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MergeRollupTaskGenerator.class);
+
+  private static final int DEFAULT_MAX_NUM_RECORDS_PER_TASK = 50_000_000;
+  private static final int DEFAULT_NUM_PARALLEL_BUCKETS = 1;
+  private static final String REFRESH = "REFRESH";
+
+  // This is the metric that keeps track of the task delay in the number of time buckets. For example, if we see this
+  // number to be 7 and merge task is configured with "bucketTimePeriod = 1d", this means that we have 7 days of
+  // delay. When operating merge/roll-up task in production, we should set the alert on this metrics to find out the
+  // delay. Setting the alert on 7 time buckets of delay would be a good starting point.
+  private static final String MERGE_ROLLUP_TASK_DELAY_IN_NUM_BUCKETS = "mergeRollupTaskDelayInNumBuckets";
+
+  // tableNameWithType -> mergeLevel -> watermarkMs
+  private Map<String, Map<String, Long>> _mergeRollupWatermarks;
+  // tableNameWithType -> maxEndTime
+  private Map<String, Long> _tableMaxEndTimeMs;
+  private ClusterInfoAccessor _clusterInfoAccessor;
+
+  @Override
+  public void init(ClusterInfoAccessor clusterInfoAccessor) {
+    _clusterInfoAccessor = clusterInfoAccessor;
+    _mergeRollupWatermarks = new HashMap<>();
+    _tableMaxEndTimeMs = new HashMap<>();
+  }
+
+  @Override
+  public String getTaskType() {
+    return MergeRollupTask.TASK_TYPE;
+  }
+
+  @Override
+  public List<PinotTaskConfig> generateTasks(List<TableConfig> tableConfigs) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 47. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.